File: android/src/main/kotlin/re/notifica/flutter/handlers/CoreHandler.kt
package re.notifica.flutter.handlers

import android.content.Context
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import re.notifica.Notificare
import re.notifica.flutter.NotificareReceiver

internal class CoreHandler(
    private val context: Context,
) : MethodChannel.MethodCallHandler {

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "isConfigured" -> getConfigured(result)
            "isReady" -> getReady(result)
            "configure" -> configure(call, result)
            "launch" -> launch(result)
            "unlaunch" -> unlaunch(result)
            else -> result.notImplemented()
        }
    }

    private fun getConfigured(result: MethodChannel.Result) {
        result.success(Notificare.isConfigured)
    }

    private fun getReady(result: MethodChannel.Result) {
        result.success(Notificare.isReady)
    }

    private fun configure(call: MethodCall, result: MethodChannel.Result) {
        val applicationKey = call.argument<String>("applicationKey")
            ?: throw IllegalArgumentException("applicationKey cannot be null")

        val applicationSecret = call.argument<String>("applicationSecret")
            ?: throw IllegalArgumentException("applicationSecret cannot be null")

        Notificare.configure(context, applicationKey, applicationSecret)
        result.success(null)
    }

    private fun launch(result: MethodChannel.Result) {
        Notificare.intentReceiver = NotificareReceiver::class.java

        Notificare.launch()
        result.success(null)
    }

    private fun unlaunch(result: MethodChannel.Result) {
        Notificare.unlaunch()
        result.success(null)
    }
}

File: ios/Classes/Handlers/CoreHandler.swift
//
//  CoreHandler.swift
//  notificare
//
//  Created by Helder Pinhal on 01/12/2020.
//

import Foundation
import NotificareSDK

class CoreHandler {

    func handler(call: FlutterMethodCall, result: FlutterResult) {
        switch call.method {
        case "isConfigured": getConfigured(call, result)
        case "isReady": getReady(call, result)
        case "configure": configure(call, result)
        case "launch": launch(call, result)
        case "unlaunch": unlaunch(call, result)
        default: result(FlutterMethodNotImplemented)
        }
    }

    private func getConfigured(_ call: FlutterMethodCall, _ result: FlutterResult) {
        result(Notificare.shared.isConfigured)
    }

    private func getReady(_ call: FlutterMethodCall, _ result: FlutterResult) {
        result(Notificare.shared.isReady)
    }

    private func configure(_ call: FlutterMethodCall, _ result: FlutterResult) {
        let arguments = call.arguments as! FlutterDictionary

        Notificare.shared.configure(
            applicationKey: arguments["applicationKey"] as! String,
            applicationSecret: arguments["applicationKey"] as! String
        )

        result(nil)
    }

    private func launch(_ call: FlutterMethodCall, _ result: FlutterResult) {
        Notificare.shared.launch()
        result(nil)
    }

    private func unlaunch(_ call: FlutterMethodCall, _ result: FlutterResult) {
        Notificare.shared.unlaunch()
        result(nil)
    }
}
